src/main/golem/util/validation/dimensions.kt

```diff
@@ -0,0 +1,256 @@
+                val actualCols = matrix.numCols()
+                if (actualRows in expectedRows && actualCols in expectedCols) {
+                    expectedRows.retainAll(arrayOf(actualRows))
+                    expectedCols.retainAll(arrayOf(actualCols))
+                } else if (actualRows in expectedCols
+                        && actualCols in expectedRows
+                        && matrix in transposable) {
+                    expectedCols.retainAll(arrayOf(actualRows))
+                    expectedRows.retainAll(arrayOf(actualCols))
+                } else {
+                    failed.add(Pair(context.matrixNames[index], matrix))
+                }
+            }
+        }
+        if (failed.size > 0) {
+            throw IndexOutOfBoundsException(concoctExceptionMessage(context, failed))
+        }
+    }
+
+
+    // Append text to the PrintStream, padded out to the given width with the given characters.
+    private fun PrintStream.writeColumn(width: Int, text: String, lpad: Char? = null, rpad: Char? = ' ',
+                                        gap: Char? = ' ') {
+        val nrpad = if (rpad == null) 0 else (width - text.length) / (if (lpad == null) 1 else 2)
+        val nlpad = if (lpad == null) 0 else (width - nrpad - text.length)
+        if (lpad != null)
+            for (x in 1..nlpad)
+                append(lpad)
+        append(text)
+        if (rpad != null)
+            for (x in 1..nrpad)
+                append(rpad)
+        if (gap != null)
+            append(gap)
+    }
+
+    // Build an english-description of numeric dimensions
+    private fun concoctFixedDescription(rows: String, cols: String): String {
+        if (rows == "1")
+            return "a ${cols}-vector"
+        if (cols == "1")
+            return "a ${rows}-vector"
+        return "${rows} by ${cols}"
+    }
+
+    // Build a list of things compareTo needs to be the same as. Looks up other things that have
+    // the same character-based dimension defined.
+    private fun concoctComparison(name: String, compareTo: String): List<String> {
+        val out = mutableListOf<String>()
+        if (compareTo !in seenInError) {
+            for ((name2, dims) in declared) {
+                val (rows, cols) = dims
+                if (name2 != name) {
+                    if (rows == compareTo) out.add("as ${name2} has rows")
+                    if (cols == compareTo) out.add("as ${name2} has columns")
+                }
+            }
+            seenInError.add(compareTo)
+        }
+        return out
+    }
+
+    // Build our pretty table of dimensions real vs actual.
+    private fun concoctExceptionMessage(context: ValidationContext, failed: List<Pair<String,
+    Matrix<Double>>>): String {
+        val strings = mutableListOf<String>()
+        for ((index, matrix) in context.matrices.withIndex()) {
+            val name = context.matrixNames[index]
+            val decl = declared[name]
```

Comment:
Nope, let is defined as inline:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/let.html

Generated bytecode is:
```kotlin
   L7
    LINENUMBER 146 L7
    ALOAD 0
    GETFIELD golem/util/validation/DimensionValidator.declared : Ljava/util/Map;
    ALOAD 8
    INVOKEINTERFACE java/util/Map.get (Ljava/lang/Object;)Ljava/lang/Object;
    CHECKCAST kotlin/Pair
    DUP
    IFNULL L8
    ASTORE 9
    NOP
   L9
    LINENUMBER 280 L9
    ALOAD 9
    CHECKCAST kotlin/Pair
    ASTORE 10
   L10
    LINENUMBER 147 L10
    ALOAD 3
    ALOAD 8
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
    POP
   L11
    LINENUMBER 148 L11
    ALOAD 10
    ASTORE 11
    ALOAD 11
    INVOKEVIRTUAL kotlin/Pair.component1 ()Ljava/lang/Object;
    CHECKCAST java/lang/String
    ASTORE 12
    ALOAD 11
    INVOKEVIRTUAL kotlin/Pair.component2 ()Ljava/lang/Object;
    CHECKCAST java/lang/String
    ASTORE 13
    ACONST_NULL
    ASTORE 11
   L12
    LINENUMBER 149 L12
    ALOAD 3
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    ALOAD 12
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    LDC "x"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 13
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
    POP
   L13
    LINENUMBER 150 L13
    ALOAD 3
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    ALOAD 7
    INVOKEINTERFACE golem/matrix/Matrix.numRows ()I
    INVOKEVIRTUAL java/lang/StringBuilder.append (I)Ljava/lang/StringBuilder;
    LDC "x"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 7
    INVOKEINTERFACE golem/matrix/Matrix.numCols ()I
    INVOKEVIRTUAL java/lang/StringBuilder.append (I)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
   L14
   L15
    INVOKESTATIC java/lang/Boolean.valueOf (Z)Ljava/lang/Boolean;
   L16
    CHECKCAST java/lang/Boolean
    GOTO L17
   L8
    POP
    ACONST_NULL
   L17
    POP
```

old:
```kotlin
    LINENUMBER 153 L18
    ALOAD 0
    GETFIELD golem/util/validation/DimensionValidator.declared : Ljava/util/Map;
    ALOAD 8
    INVOKEINTERFACE java/util/Map.get (Ljava/lang/Object;)Ljava/lang/Object;
    CHECKCAST kotlin/Pair
    ASTORE 9
   L19
    LINENUMBER 154 L19
    ALOAD 9
    IFNULL L20
   L21
    LINENUMBER 155 L21
    ALOAD 3
    ALOAD 8
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
    POP
   L22
    LINENUMBER 156 L22
    ALOAD 9
    ASTORE 12
    ALOAD 12
    INVOKEVIRTUAL kotlin/Pair.component1 ()Ljava/lang/Object;
    CHECKCAST java/lang/String
    ASTORE 10
    ALOAD 12
    INVOKEVIRTUAL kotlin/Pair.component2 ()Ljava/lang/Object;
    CHECKCAST java/lang/String
    ASTORE 11
    ACONST_NULL
    ASTORE 12
   L23
    LINENUMBER 157 L23
    ALOAD 3
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    ALOAD 10
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    LDC "x"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 11
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
    POP
   L24
    LINENUMBER 158 L24
    ALOAD 3
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    ALOAD 7
    INVOKEINTERFACE golem/matrix/Matrix.numRows ()I
    INVOKEVIRTUAL java/lang/StringBuilder.append (I)Ljava/lang/StringBuilder;
    LDC "x"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 7
    INVOKEINTERFACE golem/matrix/Matrix.numCols ()I
    INVOKEVIRTUAL java/lang/StringBuilder.append (I)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    INVOKEINTERFACE java/util/List.add (Ljava/lang/Object;)Z
    POP
   L20
    LINENUMBER 143 L20
   L25
    GOTO L2
   L3
```